radix-engine/src/engine/track.rs

```diff
@@ -518,7 +519,7 @@ impl<'s, R: FeeReserve> Track<'s, R> {
     pub fn finalize(
         mut self,
         invoke_result: Result<Vec<Vec<u8>>, RuntimeError>,
-        resource_changes: Vec<ResourceChange>, // TODO: wrong abstraction, resource change should be derived from track instead of kernel
+        execution_trace: ExecutionTrace, // TODO: wrong abstraction, resource change should be derived from track instead of kernel
```

Comment:
I'm not sure if/why the abstraction is wrong here?

Reply:
Ah, I left the comment. I was thinking that resource changes (not resource movements) should be generated through the `Track` rather that execution trace, so no need to pass the argument to `finalize`.

Reply:
If we look at all the node/substate changes made during the transaction, we should be able to tell which component has gained/lost XRD by how much.

Reply:
On the other hand, if we want to trace resource movements from X to Y, we will need to extract from execution trace.

Reply:
Hmh, maybe I'll need to refine my mental model :) I view track as being quite low-level (i.e. working with substates etc; inseparable component of the engine), while the resource changes, or the execution trace module in general, is a higher-level abstraction (ideally pluggable at some point - i.e. only enabled for tx preview), which provides some insights of the execution to the user.

Reply:
The track is doing a bit more than substates - it collects all transaction wide side effects. 

And yes some insights must be from the execution trace, especially resource flow and call stack. Track is more of a flattened view of the state/resource changes, which make it easier to be used for generating diffs.

Since this PR is out, I will review it using your model for now.